### PR TITLE
Add HiDPI support in GTK.

### PR DESCRIPTION
**Pull request by QuLogic**
## PR Summary

I had this written for GTK3 some time ago, but was waiting for GTK4 to go in so I could add that in as well. Mostly the same as Qt, but linking to the GTK version of events, etc.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/backends/backend_gtk3.py [hunk 1]
```diff
@@ -169,27 +179,26 @@ def key_release_event(self, widget, event):
     def motion_notify_event(self, widget, event):
         if event.is_hint:
             t, x, y, state = event.window.get_device_position(event.device)
+            # flipy so y=0 is bottom of canvas
+            x *= self.device_pixel_ratio
+            y = (self.get_allocation().height - y) * self.device_pixel_ratio
```
Comment (anntzer) [inline lib/matplotlib/backends/backend_gtk3.py:184]: I think you can just pass `event.window.get_device_position(event.device)` (which has `x` and `y` attributes) to `_mouse_event_coords`?

File: lib/matplotlib/backends/backend_gtk3.py [hunk 2]
```diff
@@ -211,10 +220,17 @@ def _get_key(self, event):
                     key = f'{prefix}+{key}'
         return key
 
+    def _update_device_pixel_ratio(self, instance, param):
+        # We need to be careful in cases with mixed resolution displays if
+        # device_pixel_ratio changes.
+        if self._set_device_pixel_ratio(self.get_scale_factor()):
+            self.draw()
```
Comment (anntzer) [inline lib/matplotlib/backends/backend_gtk3.py:227]: At least when testing with GDK_SCALE=2, this apparently needs to be called explicitly at some point, else the dpi is never set to 2.  Although the scaling also seems messed up even then...  (Haven't fully investigated yet, just a quick try.)

Comment (QuLogic) [reply lib/matplotlib/backends/backend_gtk3.py:227]: My GTK/GDK doesn't seem to care about `GDK_SCALE`. I originally had this called in map-event as well; maybe that's needed for you?

Comment (QuLogic) [reply lib/matplotlib/backends/backend_gtk3.py:227]: Ah, that environment variable only affects the x11 backend, not Wayland. To fix, it seems enough to trigger the pixel ratio update in `map-event`, which seems to require adding `STRUCTURE_MASK` to the event mask. Strangely, we get `configure-event` signals even though the documentation says that also needs the structure mask, so we could put the pixel ratio update there also.

This all seems to be unnecessary on Wayland, which has multi-monitor HiDPI support, and presumably had to be sure that the scale factor was correct all the time.

Comment (QuLogic) [reply lib/matplotlib/backends/backend_gtk3.py:227]: Actually, I found `screen-changed`, which seems the better signal for this.

File: lib/matplotlib/backends/backend_gtk3.py (hunk 1 above)
Comment (QuLogic) [reply lib/matplotlib/backends/backend_gtk3.py]: I removed that entirely, as that sort of event compression is deprecated.

**Comment by anntzer** [comment]
On gtk3agg/X11, I now get
```
Traceback (most recent call last):
  File ".../matplotlib/backends/backend_gtk3.py", line 234, in _screen_changed
    self._update_device_pixel_ratio()
TypeError: _update_device_pixel_ratio() missing 2 required positional arguments: 'instance' and 'param'
```

**Comment by QuLogic** [comment]
Oops, sorry, juggling multiple machines to test this, and must have got some things crossed.

What versions of GTK3 & GTK4 are you using?

**Comment by anntzer** [comment]
3.24.30/4.4.0.

**Comment by QuLogic** [comment]
Ah sorry, it's not versions; I just keep forgetting to test with the X11 backend. Agg should work now.

I totally forgot, and should have mentioned initially, that I had no idea how to get this working with Cairo. In other Cairo backends, we draw to an image surface, and blit it to the backend's widget. Unlike those backends, here we render directly to the GTK cairo context. I can fake this a bit with some scaling and translation, but the line width, text size, etc. are then smaller than they should be. But I just realized that QtCairo is wrong the same way, so this might be fine as a stopgap in the meantime.

**Comment by QuLogic** [comment]
OK, this should now "work" for Cairo as well in that the canvas is not giant and clipped. However, there is a bug across all Cairo backends with HiDPI not rendering elements at the correct size, for which I will open a separate issue.

**Comment by anntzer** [comment]
This mostly works for me now; there's only one single issue left (again, always on X11 and always on environment-variable simulated high-dpi): the initial window size on gtk3 (not on gtk4) is too large (scaled by the dpi factor, I think).

**Comment by QuLogic** [comment]
I think that might be bugginess in GTK. We get `configure-event` with already scaled up sizes in that case, whereas on a Wayland HiDPI screen, the sizes are still in logical pixels. Thus when we do `dim_in_pixels*device_pixel_ratio/dpi`, we get a doubled dimension in inches. And even if we ignore that, the window itself is already doubled in size and we'd just be drawing in a corner.

**Comment by anntzer** [comment]
To be honest that last point is a bit beyond my pay grade (or rather beyond the pay grade of my screen) but the rest looks good to me.

**Review by anntzer — APPROVED** [review]

**Comment by QuLogic** [comment]
Well, to be clear, it's the same in GTK4, the `resize` event comes in with the size in logical pixels, so that's why we don't get a double-sized window.

**Comment by QuLogic** [comment]
Hah, actually, I take that back; it is our bug. Grabbing the pixel ratio earlier meant that the DPI was updated before the window itself was created. And the initial window size calculation uses the physical size instead of the logical size.

**Review by anntzer — APPROVED** [review]
Yup, that works :)

**Review by tacaswell — APPROVED** [review]

**Comment by tacaswell** [comment]
@QuLogic can self merge on re-base and green.

**Pull request merged**